shark_turbine/kernel/wave/utils.py

```diff
@@ -59,3 +64,43 @@ def print_trace(trace: CapturedTrace, custom_print: bool = True):
         if custom_print:
             for node in subgraph.nodes:
                 print(get_custom(node))
+
+
+def remove_unused_operators(trace: CapturedTrace):
+    """
+    Removes all operators that are not used in the graph,
+    excluding output and global write nodes.
+    Repeats this process till no more operators can be removed.
+    """
+
+    def is_removable_operator(node: fx.Node) -> bool:
+        custom = get_custom(node)
+        idxc = IndexingContext.current()
+        is_global_write = (
+            isinstance(custom, Write)
+            and custom.type.address_space.subs(idxc.subs) == GLOBAL_ADDRESS_SPACE
+        )
+        return (
+            not custom.users and not isinstance(custom, Output) and not is_global_write
+        )
+
+    while removable_nodes := trace.walk(is_removable_operator):
+        for node in removable_nodes:
+            get_custom(node).graph.erase_node(node)
+
+
+def delinearize_index(index: IndexExpr, shape: list[int]) -> list[IndexExpr]:
+    """
+    Delinearizes a 1D index into a multi-dimensional index
+    based on the shapes provided. The returned array contains
+    the multi-dimensional index.
```

Comment:
NIT: Really NIT/opinion based but would be nice to have an easy sample similar to how affine dialect's delinearize_index doc:

```
%indices:3 = affine.delinearize_index %linear_index into (%c16, %c224, %c224) : index, index, index

%indices:3 conceptually holds the following:

[linear_idx floordiv 50176,  (linear_idx mod 50176) floordiv 224, linear_idx mod 224]
```